src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1378,24 +1378,27 @@ void uct_dc_mlx5_ep_handle_failure(uct_dc_mlx5_ep_t *ep, void *arg,
     ucs_arbiter_group_t *group;
     uint8_t pool_index;
 
+    ucs_debug("handle failure iface: %p, dci[%d] qpn 0x%x, status: %s", iface,
+              dci_index, txwq->super.qp_num, ucs_status_string(ep_status));
+
     ucs_assert(!uct_dc_mlx5_iface_is_dci_rand(iface));
 
     uct_dc_mlx5_update_tx_res(iface, txwq, txqp, pi);
     uct_rc_txqp_purge_outstanding(&iface->super.super, txqp, ep_status, pi, 0);
 
     ucs_assert(ep->dci != UCT_DC_MLX5_EP_NO_DCI);
-    uct_dc_mlx5_iface_dci_put(iface, dci_index);
     uct_dc_mlx5_get_arbiter_params(iface, ep, &waitq, &group, &pool_index);
 
     /* Do not invoke pending requests on a failed endpoint */
     ucs_arbiter_group_desched(waitq, group);
-    uct_dc_mlx5_iface_progress_pending(iface, pool_index);
-
     uct_dc_mlx5_iface_set_ep_failed(iface, ep, cqe, txwq, ep_status);
+    uct_dc_mlx5_iface_dci_put(iface, dci_index);
 
-    if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
-        uct_dc_mlx5_iface_reset_dci(iface, dci_index);
-    }
+    ucs_assert(ep->dci == UCT_DC_MLX5_EP_NO_DCI);
```

Comment:
this assert is failing
we should add return value to uct_dc_mlx5_iface_dci_put() to indicate if it released the dci and move lines 1397-1401 inside "if (dci_put(...) == true)"

Reply:
then reorder can be possible, if we release read FC resources in `uct_dc_mlx5_update_tx_res(iface, txwq, txqp, pi);` on line 1386 and then did not manage to release dci. In this case we will not progress pendings but may unblock other dcis/eps waiting for read fc resources.
also would need to check why the assert is failing if we just invoked purge outstanding